src/server/pfs/drive/driver.go

```diff
@@ -821,6 +821,49 @@ func (d *driver) DeleteAll(shards map[uint64]bool) error {
 	return nil
 }
 
+func (d *driver) ArchiveAll(shards map[uint64]bool) error {
+	repoInfos, err := d.ListRepo(nil, shards)
+	if err != nil {
+		return err
+	}
+	// We want to make sure we archive repos after archiving their provenance.
+	// The reason is that if a pipeline see's missing output Commits for a set of input Commits, it will trigger the pipeline, creating new output commits. This is a race.
+
+	// Provenance has a nice invariant:
+	// A in Provenance(B) => len(Provenance(A)) < len(Provenance(B))
+	// Since Provenance is transitive
+	// Thus when we sort by length we guarantee that A will have a lower index than B
+	// Because we want to make sure we delete the higher indexes first we traverse in reverse
```

Comment:
These comments are a bit out of date with the code. I'd remove them and expand to comment below to say something like: We break if the repo has Provenance because repos with provenance will already have been archived because archiving is transitive over provenance.